pkg/lib/operation/project/remote/table/upload/operation.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/schollz/progressbar/v3"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/filesystem"
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/telemetry"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+type dependencies interface {
+	KeboolaProjectAPI() *keboola.API
+	Logger() log.Logger
+	Fs() filesystem.Fs
+	Tracer() trace.Tracer
+}
+
+type Options struct {
+	TableID         keboola.TableID
+	FilePath        string
+	FileName        string
+	IncrementalLoad bool
+}
+
+func Run(ctx context.Context, o Options, d dependencies) (err error) {
+	ctx, span := d.Tracer().Start(ctx, "kac.lib.operation.project.remote.table.upload")
+	defer telemetry.EndSpan(span, &err)
+
+	d.Logger().Infof(`Checking for bucket "%s".`, o.TableID.BucketID)
+	err = ensureBucketExists(ctx, d, o.TableID.BucketID)
+	if err != nil {
+		return err
+	}
+
+	d.Logger().Infof(`Uploading file "%s".`, o.FilePath)
+	file, err := uploadFile(ctx, &o, d)
+	if err != nil {
+		return err
+	}
+	d.Logger().Infof(`Loading file "%s" into table "%s".`, o.FileName, o.TableID)
+
+	if !checkTableExists(ctx, d, o.TableID) {
+		err := d.KeboolaProjectAPI().CreateTableFromFileRequest(o.TableID, file.ID).SendOrErr(ctx)
+		if err != nil {
+			return err
+		}
+	} else {
+		job, err := d.KeboolaProjectAPI().LoadDataFromFileRequest(
+			o.TableID,
+			file.ID,
+			keboola.WithIncrementalLoad(o.IncrementalLoad),
+		).Send(ctx)
+		if err != nil {
+			return err
+		}
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+		defer cancel()
+
+		err = d.KeboolaProjectAPI().WaitForStorageJob(ctx, job)
+		if err != nil {
+			return err
+		}
+	}
+
+	d.Logger().Infof(`Uploaded "%s" to table "%s".`, o.FilePath, o.TableID)
+
+	return nil
+}
+
+func ensureBucketExists(ctx context.Context, d dependencies, id keboola.BucketID) error {
+	err := d.KeboolaProjectAPI().GetBucketRequest(id).SendOrErr(ctx)
+	var apiErr *keboola.StorageError
+	if errors.As(err, &apiErr) && apiErr.ErrCode == "storage.buckets.notFound" {
+		d.Logger().Infof(`Bucket "%s" does not exist, creating it.`, id)
+		// Bucket doesn't exist -> create it
+		bucket := &keboola.Bucket{ID: id}
+		if _, err := d.KeboolaProjectAPI().CreateBucketRequest(bucket).Send(ctx); err != nil {
+			return err
+		}
+	}
+	d.Logger().Infof(`Bucket "%s" exists.`, id)
+	return nil
+}
+
+func checkTableExists(ctx context.Context, d dependencies, id keboola.TableID) bool {
+	err := d.KeboolaProjectAPI().GetTableRequest(id).SendOrErr(ctx)
+	var apiErr *keboola.StorageError
+	if errors.As(err, &apiErr) && apiErr.ErrCode == "storage.tables.notFound" {
+		// Table doesn't exist -> create it
+		return false
+	}
+	return true
+}
+
+func uploadFile(ctx context.Context, o *Options, d dependencies) (*keboola.File, error) {
+	var reader io.Reader
+	var bar *progressbar.ProgressBar
+	{
+		file, err := d.Fs().Open(o.FilePath)
+		if err != nil {
```

Comment:
Myslel som, ze toto bude zdielat kod s `file upload` a `table import`, ktore uz mame.
Je nejaky dovod, preco je to tu riesene znova?

Reply:
>  ten kód není napsanej tak, že bych mohl jen zavolal ty dvě operace za sebou. 

Robime to bežne inde a na prvy pohlad som tam nejaky velky blocker/potrebnu upravu nenasiel.

Najprv si zozberas vsetky options, a potom zavolas za sebou tie 2 `operation`.

>Table import nevytváří table

na to je to issue (mozes si ho zobrat ak nanom nerobi uz Jakub).

> File upload nevrací ten soubor, který vytvoří.

Pridas dalsiu navratovu hodnotu [sem](https://github.com/keboola/keboola-as-code/blob/560270a37086cc27dd65d53a6abdf47cd71d3102/pkg/lib/operation/project/remote/file/upload/operation.go#L31).

> Taky jsem nechtěl řešit file ze stdin,

To je prave [vyriešené](https://github.com/keboola/keboola-as-code/blob/560270a37086cc27dd65d53a6abdf47cd71d3102/pkg/lib/operation/project/remote/file/upload/operation.go#L37), stačí to zavolať.

> Ještě by jich tady mohlo být víc (např. header-less csv, kde pak dodáš columns přes flag nebo přes dialog)

No práve, ak budeme nieco upravovat, tak to neches robit 2x.

--------------

Skusis to alebo mam poslat PR?

Reply:
Netlacim ta do toho, ze to musi byt za hodinu hotove, ale velmi nemam chut to mergut v takom stave, ze sa to moze rovno refactorovat.

Reply:
Ak by si chcel nejaky priklad zlozenej operacie, tak napr `kbc sync init`: https://github.com/keboola/keboola-as-code/blob/560270a37086cc27dd65d53a6abdf47cd71d3102/pkg/lib/operation/project/sync/init/operation.go#L57-L66

Reply:
> Najprv si zozberas vsetky options, a potom zavolas za sebou tie 2 operation.

Vysvětlil jsem, proč to nejde. Musel bych to víc upravit. Ještě dodám že ten command by musel ty options podporovat nebo je nějak konvertovat tak, aby ty operace fungovaly, což taky není otázka pár minut.

> Netlacim ta do toho, ze to musi byt za hodinu hotove, ale velmi nemam chut to mergut v takom stave, ze sa to moze rovno refactorovat.

Tak proč se mergnuly Jakubovy commandy, když nejsou dokončené a musí se zrefaktorovat?

Reply:
Lebo boli funkcne a isiel na dovolenku. 
Nemusia sa zrefaktorovat, musia sa doplnit.

Ak by nesiel na dovolenku, tak nezalozim [nove tasky,](https://keboola.atlassian.net/browse/PSGO-131) ale riesime to v PR (a iba na nieco by sa asi zalozil task).
Nemal tam ziaden duplicitny kod, len bola potrebna dalsia iteracia, na ktoru mu pred dovolenkou nezostal cas.

Myslim, ze sa vobec nemusime o tom natahovat :slightly_smiling_face: , staci povedat, ze to neplanujes upravit, poslem PR na doplnenie.

Reply:
Ak chces presne ten isty postup, tak to mergnime, zaloz novy task `table upload improvements` a upravy daj ako novy PR, nemam s tym ziadny problem.